### PR TITLE
net: coap_client: fix incorrect handling of responses longer than CONFIG_COAP_CLIENT_MESSAGE_SIZE

### LINKED ISSUE #76089: net: coap_client: truncation of packets longer than CONFIG_COAP_CLIENT_MESSAGE_SIZE
**Describe the bug**
When making requests using `coap_client_req`, responses longer than `CONFIG_COAP_CLIENT_MESSAGE_SIZE` can be truncated, or result in incorrect handling of the subsequent block transfer (depending on the total length of the resource).

This issue can manifest itself in three different ways (that I can see), but all are related:

1) Response is longer than `CONFIG_COAP_CLIENT_MESSAGE_SIZE` but shorter than `MAX_COAP_MSG_LEN` (because the CoAP header is shorter than the maximum value).

In this case no truncation happens, the whole payload is delivered to the application callback. This is OK although perhaps slightly counter-intuitive: the application developer might have assumed that the callback can't deliver a message payload longer than `CONFIG_COAP_CLIENT_MESSAGE_SIZE` and could have sized downstream buffers etc. accordingly.

For example with `CONFIG_COAP_CLIENT_MESSAGE_SIZE=256` and a resource length of 270 bytes:

```
[00:00:00.110,000] <inf> coap_download: Starting CoAP download using IPv6
[00:00:00.180,000] <inf> coap_download: CoAP response, result_code=69, offset=0, len=270
[00:00:00.180,000] <inf> coap_download: CoAP download done, got 270 bytes in 70 ms
```

2) Response is longer than `CONFIG_COAP_CLIENT_MESSAGE_SIZE`, longer than `MAX_COAP_MSG_LEN`, but from the server's point of view still fits in a single block (because the server defaults to a larger block size than the client).

In this case the application receives a single callback with a truncated response, and no indication that the response is truncated. The CoAP client doesn't send any further block requests to retrieve the rest of the data.

For example with `CONFIG_COAP_CLIENT_MESSAGE_SIZE=256` and a resource length of 512 bytes:

```
[00:00:00.110,000] <inf> coap_download: Starting CoAP download using IPv6
[00:00:00.180,000] <inf> coap_download: CoAP response, result_code=69, offset=0, len=279
[00:00:00.180,000] <inf> coap_download: CoAP download done, got 279 bytes in 70 ms
```

But in wireshark, we can see that the server sent a payload of 512 bytes:
image

3) Response is longer than `CONFIG_COAP_CLIENT_MESSAGE_SIZE`, longer than `MAX_COAP_MSG_LEN`, and longer than the server's default block size.

In this case, the CoAP client receives the first response, parses the Block2 option sent by the server, and sends requests for the remaining blocks that it is missing. But because the first packet was longer than `CONFIG_COAP_CLIENT_MESSAGE_SIZE`, the offsets passed to the application callback for all subsequent blocks are wrong - they are not a multiple of the block size. So it appears to the application as though more data has been received than the actual size of the resource (since some data has been received twice).

For example with `CONFIG_COAP_CLIENT_MESSAGE_SIZE=256` and a resource length of 2048 bytes:

```
[00:00:00.110,000] <inf> coap_download: Starting CoAP download using IPv6
[00:00:00.180,000] <inf> coap_download: CoAP response, result_code=69, offset=0, len=279
[00:00:00.240,000] <inf> coap_download: CoAP response, result_code=69, offset=279, len=256
[00:00:00.300,000] <inf> coap_download: CoAP response, result_code=69, offset=535, len=256
[00:00:00.360,000] <inf> coap_download: CoAP response, result_code=69, offset=791, len=256
[00:00:00.420,000] <inf> coap_download: CoAP response, result_code=69, offset=1047, len=256
[00:00:00.480,000] <inf> coap_download: CoAP response, result_code=69, offset=1303, len=256
[00:00:00.540,000] <inf> coap_download: CoAP response, result_code=69, offset=1559, len=256
[00:00:00.600,000] <inf> coap_download: CoAP response, result_code=69, offset=1815, len=256
[00:00:00.600,000] <inf> coap_download: CoAP download done, got 2071 bytes in 490 ms
```

and in Wireshark:
image

**To Reproduce**
Check out the PR: https://github.com/zephyrproject-rtos/zephyr/pull/76069

Build and run using the instructions in the `README.rst`, using `native_sim` is easiest.

Change the size of the resource sent by the server to see each of the different issues observed above.

**Expected behavior**
Any data that can't fit in the client's receive buffer should be retrieved by subsequent block transfer requests. The offset of each block passed to the application should be correct (ie. a multiple of the block size).

**Impact**
The application can't reliably receive CoAP data of any arbitrary length, unless the server's default block size happens to be less than or equal to the client's max message size.

**Environment (please complete the following information):**

 - OS: Linux
 - Toolchain: Zephyr SDK
 - Commit SHA or Version used: latest `main` at the time of writing (`b8ae3a416b9`).

**Additional context**

I'd be happy to help with a fix for this if that's useful, I can see a few changes that might be needed but could do with some guidance on the best way to go:

1) Use the `MSG_TRUNC` flag when receiving from the socket, so that we know the true length of the packet and whether it was truncated. Use this info when deciding if a block transfer is needed, not just the Block2 option in the response packet.
2) Ensure that the block offset is only ever incremented by a multiple of the block size
3) (Not necessary but might be helpful): always send a Block2 option in the first request packet, to ensure that the server doesn't respond with more data than the client can handle - ie. `Figure 3: Block-Wise GET with Early Negotiation` from https://datatracker.ietf.org/doc/html/rfc7959#section-3.1

### PR DESCRIPTION
**Pull request by mrodgers-witekio**
Resolves https://github.com/zephyrproject-rtos/zephyr/issues/76089

Add the `MSG_TRUNC` flag to the `recv` options, so that the CoAP client is aware when the received UDP packet has been truncated. This info is used along with the presence of a Block2 option in the response to determine whether a blockwise transfer is needed to fetch the rest of the resource.

Additionally the size of the CoAP payload passed to the application callback is limited to the block size if a blockwise transfer is in progress. This avoids any overlapping data being sent to the application in the event that the first response of a blockwise transfer uses a larger block size than the client can handle, and therefore contains some data that is also part of the following block.

The behaviour now is:
1) Response is longer than `CONFIG_COAP_CLIENT_MESSAGE_SIZE` but still fits into the receive buffer with no truncation --> no change to behaviour, whole payload delivered to application callback.
2) Response is longer than `CONFIG_COAP_CLIENT_MESSAGE_SIZE` and longer than the CoAP client's receive buffer --> Only the first `CONFIG_COAP_CLIENT_MESSAGE_SIZE` bytes are delivered to the application callback, and the remaining data is fetched via blockwise transfer.

### CODE REVIEW

File: subsys/net/lib/coap/coap_client.c [hunk 1]
```diff
@@ -541,30 +541,32 @@ static bool token_compare(struct coap_client_internal_request *internal_req,
...
 		LOG_ERR("Error reading response: %d", errno);
 		return -EINVAL;
-	} else if (len == 0) {
+	} else if (total_len == 0) {
 		LOG_ERR("Zero length recv");
 		return -EINVAL;
 	}
 
-	LOG_DBG("Received %d bytes", len);
+	available_len = MIN(total_len, MAX_COAP_MSG_LEN);
```
Comment (rlubos) [inline subsys/net/lib/coap/coap_client.c:560]: Technically the same, but I'd be easier to follow if `sizeof(client->recv_buf)` was used here.

File: subsys/net/lib/coap/coap_client.c [hunk 2]
```diff
@@ -896,7 +907,9 @@ void coap_client_recv(void *coap_cl, void *a, void *b)
 					continue;
 				}
 
-				ret = handle_response(clients[i], &response);
+				response_truncated = ret > MAX_COAP_MSG_LEN;
```
Comment (rlubos) [inline subsys/net/lib/coap/coap_client.c:910]: I think it'd make sense if `recv_response()` accepted additional `bool *truncated` parameter, and fill out this information.

**Review by rlubos — COMMENTED** [review]
Looks quite straightforward, although I'm a bit concerned if such behavior is spec compliant (I couldn't find an answer by briefly looking through rfc7959). From the server perspective, we initiate a block2 transfer with a block number != 0.

**Comment by mrodgers-witekio** [comment]
> Looks quite straightforward, although I'm a bit concerned if such behavior is spec compliant (I couldn't find an answer by briefly looking through rfc7959). From the server perspective, we initiate a block2 transfer with a block number != 0.

I think what's implemented here is quite similar to this example from RFC 7959:

```
   In the third example (Figure 4), the client is surprised by the need
   for a block-wise transfer, and unhappy with the size chosen
   unilaterally by the server.  As it did not send a size proposal
   initially, the negotiation only influences the size from the second
   message exchange onward.  Since the client already obtained both the
   first and second 64-byte block in the first 128-byte exchange, it
   goes on requesting the third 64-byte block ("2/0/64").  None of this
   is (or needs to be) understood by the server, which simply responds
   to the requests as it best can.

   CLIENT                                                     SERVER
     |                                                          |
     | CON [MID=1234], GET, /status                     ------> |
     |                                                          |
     | <------   ACK [MID=1234], 2.05 Content, 2:0/1/128        |
     |                                                          |
     | CON [MID=1235], GET, /status, 2:2/0/64           ------> |
     |                                                          |
     | <------   ACK [MID=1235], 2.05 Content, 2:2/1/64         |
     |                                                          |
     | CON [MID=1236], GET, /status, 2:3/0/64           ------> |
     |                                                          |
     | <------   ACK [MID=1236], 2.05 Content, 2:3/1/64         |
     |                                                          |
     | CON [MID=1237], GET, /status, 2:4/0/64           ------> |
     |                                                          |
     | <------   ACK [MID=1237], 2.05 Content, 2:4/1/64         |
     |                                                          |
     | CON [MID=1238], GET, /status, 2:5/0/64           ------> |
     |                                                          |
     | <------   ACK [MID=1238], 2.05 Content, 2:5/0/64         |

              Figure 4: Block-Wise GET with Late Negotiation
```

The only difference being that the server did not necessarily send a Block2 option in the first request.

There's also this from the Introduction section of the RFC:

```
   o  If needed, the Block options can also be used (without changes) to
      provide random access to power-of-two sized blocks within a
      resource representation.
```

Which would definitely cover this use-case (and more).

**Review by SeppoTakalo — DISMISSED** [review]

**Comment by SeppoTakalo** [comment]
@mrodgers-witekio 
Since you are working on this one, can you also check if we can do following in the `coap_client_request()`

```diff
  diff --git a/subsys/net/lib/coap/coap_client.c b/subsys/net/lib/coap/coap_client.c
  index 96d1584ccb0..17dc227532a 100644
  --- a/subsys/net/lib/coap/coap_client.c
  +++ b/subsys/net/lib/coap/coap_client.c
  @@ -148,6 +148,7 @@ static int coap_client_init_request(struct coap_client *client,
   {
          int ret = 0;
          int i;
  +       bool block2 = false;
   
          memset(client->send_buf, 0, sizeof(client->send_buf));
   
  @@ -189,6 +190,7 @@ static int coap_client_init_request(struct coap_client *client,
   
          /* Blockwise receive ongoing, request next block. */
          if (internal_req->recv_blk_ctx.current > 0) {
  +               block2 = true;
                  ret = coap_append_block2_option(&internal_req->request,
                                                  &internal_req->recv_blk_ctx);
   
  @@ -200,6 +202,9 @@ static int coap_client_init_request(struct coap_client *client,
   
          /* Add extra options if any */
          for (i = 0; i < req->num_options; i++) {
  +               if (block2 && req->options[i].code == COAP_OPTION_BLOCK2) {
  +                       continue;
  +               }
                  ret = coap_packet_append_option(&internal_req->request, req->options[i].code,
                                                  req->options[i].value, req->options[i].len);
   
```

This would allow application to add BLOCK2 option into the first query already. Then following queries should not append the same option again as the next calculated BLOCK2 options is already added.

**Review by rlubos — DISMISSED** [review]

**Comment by fabiobaltieri** [comment]
merge conflict, can you rebase please?

**Review by SeppoTakalo — APPROVED** [review]

**Review by fabiobaltieri — APPROVED** [review]

**Review by rlubos — APPROVED** [review]

**Comment by mrodgers-witekio** [comment]
Rebased again now that CI tests should be working

**Pull request merged**